File: release_report_update/release_report.go
## to run this:
# go run release_report.go

package main

import (
	"encoding/json"
	"fmt"
	"html/template"
	"io/ioutil"
	"net/http"
	"os"
	"time"
)

// Struct for service configuration
type Service struct {
	Service string `json:"service"`
	Repo    string `json:"repo"`
}

// Struct for GitHub commit data
type Commit struct {
	SHA     string `json:"sha"`
	Message string `json:"commit"`
	URL     string `json:"html_url"`
	Date    string `json:"date"`
}

// Load service configuration from config.json
func loadConfig(filename string) ([]Service, error) {
	file, err := ioutil.ReadFile(filename)
	if err != nil {
		return nil, err
	}

	var config struct {
		Services []Service `json:"services"`
	}
	err = json.Unmarshal(file, &config)
	return config.Services, err
}

// Fetch commits from GitHub API
func fetchGithubCommits(repo string, startDate, endDate string) ([]Commit, error) {
	url := fmt.Sprintf("https://api.github.com/repos/%s/commits?since=%s&until=%s", repo, startDate, endDate)
	req, _ := http.NewRequest("GET", url, nil)
	req.Header.Set("Authorization", "token YOUR_GITHUB_TOKEN")

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var commits []Commit
	err = json.NewDecoder(resp.Body).Decode(&commits)
	return commits, err
}

// Generate and save the HTML report
func generateHTMLReport(services []Service, startDate, endDate string) {
	const templateHTML = `
<!DOCTYPE html>
<html>
<head>
	<title>Release Report</title>
	<style>
		body { font-family: Arial, sans-serif; margin: 20px; }
		.container { max-width: 900px; margin: auto; background: white; padding: 20px; }
		.section { margin-bottom: 30px; }
		.service { font-weight: bold; color: #0073e6; }
		.commit { color: #ff9800; }
		.commit-link { text-decoration: none; color: #0073e6; }
	</style>
</head>
<body>
	<div class="container">
		<h1>🚀 Release Report - {{.Date}}</h1>

		<!-- Summary Report by Environment -->
		<div class="section">
			<h2>📌 Summary Report by Environment</h2>
			{{range .Services}}
				<h3 class="service">{{.Service}}</h3>
				<ul>
				{{range .Commits}}
					<li class="commit"><a href="{{.URL}}" class="commit-link">{{.Message}}</a> - {{.Date}}</li>
				{{end}}
				</ul>
			{{end}}
		</div>
	</div>
</body>
</html>`

	reportFile, err := os.Create("release_report.html")
	if err != nil {
		fmt.Println("Error creating HTML file:", err)
		return
	}
	defer reportFile.Close()

	tmpl, _ := template.New("report").Parse(templateHTML)
	reportData := struct {
		Date     string
		Services []struct {
			Service string
			Commits []Commit
		}
	}{
		Date:     time.Now().Format("January 2, 2006"),
		Services: []struct {
			Service string
			Commits []Commit
		}{},
	}

	for _, service := range services {
		commits, _ := fetchGithubCommits(service.Repo, startDate, endDate)
		reportData.Services = append(reportData.Services, struct {
			Service string
			Commits []Commit
		}{service.Service, commits})
	}

	tmpl.Execute(reportFile, reportData)
	fmt.Println("✅ HTML Release Report generated successfully!")
}

// Main function to execute the report generation
func main() {
	services, err := loadConfig("config.json")
	if err != nil {
		fmt.Println("Error loading config:", err)
		return
	}

	// Default date range (last 14 days)
	startDate := time.Now().AddDate(0, 0, -14).Format("2006-01-02")
	endDate := time.Now().Format("2006-01-02")

	// Allow user input for date range
	fmt.Printf("Enter start date (YYYY-MM-DD) [Default: %s]: ", startDate)
	fmt.Scanln(&startDate)
	fmt.Printf("Enter end date (YYYY-MM-DD) [Default: %s]: ", endDate)
	fmt.Scanln(&endDate)

	// Generate the report
	generateHTMLReport(services, startDate, endDate)
}

File: release_report_update/release_report.py
import requests
import json
import boto3
import datetime

# Load configuration from JSON file
with open("config.json") as config_file:
    config = json.load(config_file)

# Default date range for commit history (Last 14 days)
DEFAULT_START_DATE = (datetime.datetime.utcnow() - datetime.timedelta(days=14)).strftime("%Y-%m-%d")
DEFAULT_END_DATE = datetime.datetime.utcnow().strftime("%Y-%m-%d")

# User input for custom date range
start_date = input(f"Enter start date (YYYY-MM-DD) [Default: {DEFAULT_START_DATE}]: ") or DEFAULT_START_DATE
end_date = input(f"Enter end date (YYYY-MM-DD) [Default: {DEFAULT_END_DATE}]: ") or DEFAULT_END_DATE

# Function: Fetch recent commits from GitHub API for a given repository
def fetch_github_commits(repo):
    """Retrieve commit history for a specific GitHub repository."""
    url = f"https://api.github.com/repos/{repo}/commits?since={start_date}&until={end_date}"
    headers = {"Authorization": f"token YOUR_GITHUB_TOKEN"}

    response = requests.get(url, headers=headers)

    if response.status_code == 200:
        return response.json()  # Return commit data as JSON
    else:
        print(f"Error fetching commits for {repo}: {response.status_code}")
        return []

# Function: Fetch deployment details from AWS ECS
def fetch_aws_deployments():
    """Retrieve ECS cluster and service deployments from AWS."""
    session = boto3.Session(aws_access_key_id="YOUR_AWS_ACCESS_KEY", aws_secret_access_key="YOUR_AWS_SECRET_KEY")
    ecs_client = session.client("ecs")

    clusters = ecs_client.list_clusters()["clusterArns"]
    deployments = []

    for cluster in clusters:
        services = ecs_client.list_services(cluster=cluster)["serviceArns"]
        deployments.append({"cluster": cluster, "services": services})

    return deployments

# Function: Generate and save the HTML report
def generate_html_report():
    """Create a structured HTML report with summary, details, and commit history."""

    report = f"""
    <!DOCTYPE html>
    <html>
    <head>
        <title>Release Report</title>
        <style>
            body {{ font-family: Arial, sans-serif; background: #f4f4f4; margin: 20px; }}
            .container {{ max-width: 900px; margin: auto; background: white; padding: 20px; border-radius: 10px; box-shadow: 0 0 10px rgba(0,0,0,0.1); }}
            h1, h2 {{ text-align: center; }}
            .section {{ margin-bottom: 20px; }}
            .env {{ background: #e6f7ff; padding: 10px; border-left: 5px solid #0073e6; margin-bottom: 10px; }}
            .commit-link {{ color: #0073e6; text-decoration: none; }}
            .commit-link:hover {{ text-decoration: underline; }}
        </style>
    </head>
    <body>
    <div class="container">
        <h1>🚀 Release Report - {datetime.datetime.utcnow().strftime('%Y-%m-%d')}</h1>
    """

    # Summary Report by Environment
    report += "<div class='section'><h2>📌 Summary Report by Environment</h2>"
    for service in config["services"]:
        commits = fetch_github_commits(service["repo"])
        latest_commit = commits[0]["commit"]["message"] if commits else "No commits found"

        report += f"<div class='env'><h3>{service['service']}</h3><ul>"
        report += f"<li>{latest_commit} (<a href='{commits[0]['html_url']}' class='commit-link'>View</a>)</li>" if commits else "<li>No commits found</li>"
        report += "</ul></div>"
    report += "</div>"

    # Detailed Report by Service
    report += "<div class='section'><h2>🛠️ Detailed Report by Service</h2>"
    deployments = fetch_aws_deployments()

    for deployment in deployments:
        report += f"<div class='env'><h3>{deployment['cluster']}</h3><ul>"
        for service in deployment["services"]:
            report += f"<li>Service: {service}</li>"
        report += "</ul></div>"
    report += "</div>"

    # Commit History Report
    report += "<div class='section'><h2>📜 Commit History Report</h2>"
    for service in config["services"]:
        commits = fetch_github_commits(service["repo"])

        report += f"<div class='env'><h3>{service['service']}</h3><ul>"
        for commit in commits:
            report += f"<li><a href='{commit['html_url']}' class='commit-link'>{commit['commit']['message']} ({commit['commit']['committer']['date']})</a></li>"
        report += "</ul></div>"
    report += "</div>"

    report += "</div></body></html>"

    # Save the report as an HTML file
    with open("release_report.html", "w") as file:
        file.write(report)

    print("✅ HTML Release Report generated successfully!")

# Execute the report generation process
generate_html_report()
